pkg/gardenlet/controller/shoot/shoot_care_control.go

```diff
@@ -376,6 +367,12 @@ func (r *careReconciler) care(ctx context.Context, gardenClientSet kubernetes.In
 	return nil
 }
 
+func buildShootConditions(shootConditions []gardencorev1beta1.Condition, conditions []gardencorev1beta1.Condition, conditionTypes []gardencorev1beta1.ConditionType) []gardencorev1beta1.Condition {
+	result := gardencorev1beta1helper.RemoveConditions(shootConditions, conditionTypes...)
```

Comment:
Can you explain, why you remove the conditions before merging?

Reply:
Because what we should do here is update only the "shoot" conditions, not all conditions any more (since there is now another controller taking care of the "seed" conditions). Both controllers must take care to correctly update "their" conditions and not touch the conditions of the other one. Both should therefore do the following:
* Remove all conditions they care about from the list of existing conditions (in case they are no longer needed)
* Merge the calculated new conditions into the list

I kept the above logic intact until I realised that the shoot care controller was removing the seed conditions added by the GCM shoot controller (the 2 controllers were basically racing adding and removing these conditions). I then added the merge with existing conditions, but then the unit test started failing, since this allowed shoot conditions that were not longer needed to stay in the list. This helped me understand that we need the remove + merge logic also here. If there were not comprehensive unit tests, I would have easily missed this case as well.

Reply:
Ah thanks for explaining. That makes perfect sense, wonder why I didn't come to this conclusion right away.
Can you add a small comment explaining this?

Reply:
Added a comment as requested.